lib/ruby_lsp/listeners/document_symbol.rb

```diff
@@ -98,6 +106,91 @@ def on_constant_write_node_enter(node)
         )
       end
 
+      sig { params(node: Prism::ConstantPathAndWriteNode).void }
+      def on_constant_path_and_write_node_enter(node)
+        create_document_symbol(
+          name: node.target.location.slice,
+          kind: Constant::SymbolKind::CONSTANT,
+          range_location: node.location,
+          selection_range_location: node.target.location,
+        )
+      end
+
+      sig { params(node: Prism::ConstantPathOrWriteNode).void }
+      def on_constant_path_or_write_node_enter(node)
+        create_document_symbol(
+          name: node.target.location.slice,
+          kind: Constant::SymbolKind::CONSTANT,
+          range_location: node.location,
+          selection_range_location: node.target.location,
+        )
+      end
+
+      sig { params(node: Prism::ConstantPathOperatorWriteNode).void }
+      def on_constant_path_operator_write_node_enter(node)
+        create_document_symbol(
+          name: node.target.location.slice,
+          kind: Constant::SymbolKind::CONSTANT,
+          range_location: node.location,
+          selection_range_location: node.target.location,
+        )
+      end
+
+      sig { params(node: Prism::ConstantOrWriteNode).void }
+      def on_constant_or_write_node_enter(node)
+        create_document_symbol(
+          name: node.name.to_s,
+          kind: Constant::SymbolKind::CONSTANT,
+          range_location: node.location,
+          selection_range_location: node.name_loc,
+        )
+      end
+
+      sig { params(node: Prism::ConstantAndWriteNode).void }
+      def on_constant_and_write_node_enter(node)
+        create_document_symbol(
+          name: node.name.to_s,
+          kind: Constant::SymbolKind::CONSTANT,
+          range_location: node.location,
+          selection_range_location: node.name_loc,
+        )
+      end
+
+      sig { params(node: Prism::ConstantOperatorWriteNode).void }
+      def on_constant_operator_write_node_enter(node)
+        create_document_symbol(
+          name: node.name.to_s,
+          kind: Constant::SymbolKind::CONSTANT,
+          range_location: node.location,
+          selection_range_location: node.name_loc,
+        )
+      end
+
+      sig { params(node: Prism::ConstantTargetNode).void }
+      def on_constant_target_node_enter(node)
+        create_document_symbol(
+          name: node.name.to_s,
+          kind: Constant::SymbolKind::CONSTANT,
+          range_location: node.location,
+          selection_range_location: node.location,
+        )
+      end
+
+      sig { params(node: Prism::ConstantPathTargetNode).void }
+      def on_constant_path_target_node_enter(node)
+        name = constant_name(node)
+        return unless name
+
+        create_document_symbol(
+          name: name,
+          kind: Constant::SymbolKind::CONSTANT,
+          range_location: node.location,
+          selection_range_location: node.location,
+        )
+      rescue NoMethodError
+        nil
+      end
```

Comment:
I'll try to rephrase, sorry.

- in `self::Foo = 42`, this is a `ConstantPathWriteNode`, we are invoking `node.target.location.slice`, it work.
- in `self::Foo, BAR = 42`, we have this AST: 

```
@ MultiWriteNode (location: (27,0)-(27,19))
├── lefts: (length: 2)
│   ├── @ ConstantPathTargetNode (location: (27,0)-(27,9))
│   │   ├── parent:
│   │   │   @ SelfNode (location: (27,0)-(27,4))
│   │   ├── child:
│   │   │   @ ConstantReadNode (location: (27,6)-(27,9))
│   │   │   └── name: :Foo
│   │   └── delimiter_loc: (27,4)-(27,6) = "::"
│   └── @ ConstantTargetNode (location: (27,11)-(27,14))
│       └── name: :BAR
├── rest: ∅
├── rights: (length: 0)
├── lparen_loc: ∅
├── rparen_loc: ∅
├── operator_loc: (27,15)-(27,16) = "="
└── value:
    @ IntegerNode (location: (27,17)-(27,19))
    └── flags: decimal
```

When we handle `self::Foo` this is a `ConstantPathTargetNode`, [`node.full_name`](https://github.com/ruby/prism/blob/d950a865230cf29633301439b58aebf81c352767/lib/prism/node_ext.rb#L134-L145) invoke `parent&.full_name_parts`, where `full_name_parts` method does not exist on a `SelfNode`, or `CallNode`.

All three provided example will raise a `NoMethodError`. To reproduce, you can just remove this rescue, and run `bin/test test/requests/document_symbol_expectations_test.rb`, it will raises error like this one: 

```
DocumentSymbolExpectationsTest#test_document_symbol__const:
NoMethodError: undefined method `full_name_parts' for @ SelfNode (location: (27,0)-(27,4))
:Prism::SelfNode
```

I hope this is more clear, let me know. 🙏